File: python/src/causentra/relationships.py
"""Portable multi-agent relationship vocabulary."""

from __future__ import annotations

from typing import Literal

from .types import EventAttributes

RelationshipKind = Literal["handoff", "delegation"]


def relationship_attributes(
    kind: RelationshipKind,
    from_agent: str,
    to_agent: str,
    relationship_id: str | None = None,
) -> EventAttributes:
    """Build framework-neutral handoff or delegation attributes."""

    for value, field in ((from_agent, "from_agent"), (to_agent, "to_agent")):
        if not value.strip() or len(value) > 256:
            raise ValueError(f"{field} must be non-empty and at most 256 characters")
    if relationship_id is not None and (not relationship_id or len(relationship_id) > 256):
        raise ValueError("relationship_id must be non-empty and at most 256 characters")
    result: EventAttributes = {
        "causentra.agent.relationship.kind": kind,
        "causentra.agent.from.name": from_agent,
        "causentra.agent.to.name": to_agent,
    }
    if relationship_id is not None:
        result["causentra.agent.relationship.id"] = relationship_id
    return result

File: packages/sdk/src/relationships.ts
import type { EventAttributes } from "./types.js";

/** Portable relationship between two logical agents. */
export type AgentRelationshipKind = "handoff" | "delegation";

export interface AgentRelationship {
  readonly kind: AgentRelationshipKind;
  readonly fromAgent: string;
  readonly toAgent: string;
  /** Framework or application correlation ID; content must not be placed here. */
  readonly relationshipId?: string;
}

/** Builds validated, framework-neutral handoff/delegation attributes. */
export function createAgentRelationshipAttributes(
  relationship: AgentRelationship,
): EventAttributes {
  name(relationship.fromAgent, "fromAgent");
  name(relationship.toAgent, "toAgent");
  if (
    relationship.relationshipId !== undefined
    && (relationship.relationshipId.length === 0 || relationship.relationshipId.length > 256)
  ) {
    throw new TypeError("relationshipId must be a non-empty string up to 256 characters");
  }
  return {
    "causentra.agent.relationship.kind": relationship.kind,
    "causentra.agent.from.name": relationship.fromAgent,
    "causentra.agent.to.name": relationship.toAgent,
    ...(relationship.relationshipId === undefined
      ? {}
      : { "causentra.agent.relationship.id": relationship.relationshipId }),
  };
}

function name(value: string, field: string): void {
  if (value.trim().length === 0 || value.length > 256) {
    throw new TypeError(`${field} must be a non-empty string up to 256 characters`);
  }
}
